src/rdiff_backup/compare.py

```diff
@@ -103,13 +104,27 @@ def Verify(mirror_rp, inc_rp, verify_time):
                 "doesn't match recorded digest of\n   %s\n"
                 "Your backup repository may be corrupted!" %
                 (repo_rorp.get_safeindexpath(), computed_hash,
-                 repo_rorp.get_sha1()), 2)
+                 verify_sha1), 2)
     RepoSide.close_rf_cache()
     if not bad_files:
         log.Log("Every file verified successfully.", 3)
     return bad_files
 
 
+def get_hash(repo_rorp):
+    """ Try to get a sha1 digest from the repository.  If hardlinks
+    are saved in the metadata, get the sha1 from the first hardlink """
+    Hardlink.add_rorp(repo_rorp)
```

Comment:
I see a potential issue with the fact that (and I know it's not your code, and I don't understand everything yet):

- add_rorp adds a hardlink if it doesn't exist but does NOT increment the "remaining numlinks" if it already exists
- whereas del_rorp does decrement it in all cases (and potentially remove it)

(sorry, still struggling with this part of the code)

Reply:
Every hardlinked RORP is processed with `add_rorp` and then later with `del_rorp`.  When a hardlinked RORP is first added with `add_rorp`, its inode is added to `_inode_index` along with the total number of hardlinks for the inode.  Additional calls to `add_rorp` for other RORPs linked to that inode do not change the hardlink count.  Calls to `del_rorp` will decrement the hardlink count, and once zero, the inode is removed from `_inode_index`.  At that point, all of the RORPs that were hardlinked to the inode will have been seen and processed, and we can be sure that we will not encounter any other RORPs linked to that inode.

In theory, if a backup does not include _all_ of the files that are hardlinked to an inode, then the hardlink count will never decrement completely to zero.  That is not a serious issue.  It only means the inode will never be removed from the `_inode_index`, and so the memory that it uses can never be freed before the the backup is finished.

(In practice and when possible, users would do well to include all the hardlinks for a given inode in their backups.  Doing otherwise would mean that, when restored, those files will no longer be hardlinked the same way as when they were backed up.)

You would not want the code to simply increment and decrement the inode hardlink count as calls are made with the same inode to `add_rorp` and `del_rorp`, respectively.  You must keep the information about a hardlinked inode in memory for as long as there is a possibly of encountering that same inode again later in the backup process.

I'll also mention that this PR eliminates all of these warnings mentioned in issue #78:
```
Warning: Cannot find SHA1 digest for file 'file1_linked',
perhaps because this feature was added in v1.1.1
```
(I expect that I will eventually be submitting a PR with a suggested fix for the failed unit test mentioned in #78.)

Reply:
Thanks for the explanations. Makes sense. Let me continue the review, it's quite a lot I need to swallow.